File: Problems/0155_Min_Stack/Min_Stack.cs
using System;

public class MinStack {

    private int[] stack = new int[65535];
    private int index = 0;

    /** initialize your data structure here. */
    public MinStack() {
    }
    
    public void Push(int x) {
        stack[index++] = x;
        Console.WriteLine("Push(" + x.ToString() + ");");
    }
    
    public void Pop() {
        --index;
        Console.WriteLine("Pop();");
    }
    
    public int Top() {
        return stack[index - 1];
        Console.WriteLine("Top();");
    }
    
    public int GetMin() {
        int min_val = stack[index - 1];

        for (int i = 0; i < index - 1; ++i){
            if (stack[i] < min_val) {
                min_val = stack[i];
            }
        }

        Console.WriteLine("GetMin() => " + min_val.ToString());
        return min_val;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(x);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */
public class Solution
{
    private void calc()
    {
        MinStack minStack = new MinStack();
        minStack.Push(-2);
        minStack.Push(0);
        minStack.Push(-3);
        minStack.GetMin();
        minStack.Pop();
        minStack.Top();
        minStack.GetMin();
    }

    public void Main(string args)
    {
        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        calc();

        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0100_0199/0155_Min_Stack/Project_Python3/Min_Stack.py
import sys

class Node:
    def __init__(self, data):
        self.next = None
        self.data = data

class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min = sys.maxsize
        self.head = None

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x < self.min:
            self.min = x
        
        self.node = Node(self.min)
        self.node.next = Node(x)

        if self.head == None or self.head.next == None:
            self.head = self.node
        else:
            temp = self.head
            self.head = self.node
            self.head.next.next = temp

    def pop(self):
        """
        :rtype: void
        """
        if self.head == None:
            dummy = 0
        else:
            currHead = self.head.next
            if currHead.next != None:
                self.head = currHead.next
                self.min = self.head.data
            else:
                self.head = None
                self.min = sys.maxsize

    def top(self):
        """
        :rtype: int
        """
        if self.head != None:
            return self.head.next.data
        else:
            return 0

    def getMin(self):
        """
        :rtype: int
        """
        return self.min

def calc():
    minStack = MinStack()

    minStack.push(-2)
    print("push(-2)")

    minStack.push(0)
    print("push(0)")

    minStack.push(-3)
    print("push(-3)")

    result = minStack.getMin()
    print("getMin() => {0}".format(result))

    minStack.pop()
    print("pop()")

    result = minStack.top()
    print("top() => {0}".format(result))

    result = minStack.getMin()
    print("getMin() => {0}".format(result))

def main():
    calc()

if __name__ == "__main__":
    main()
